parquet-variant/src/main/java/org/apache/parquet/variant/VariantConverters.java

```diff
@@ -29,755 +29,720 @@
+                                                   ParentConverter<VariantBuilder> parent) {
+    ValueConverter converter = newValueConverter(variantGroup, parent);
 
-    // The following are only used if this is an object field.
-    private String objectFieldName = null;
-    private VariantConverter parent = null;
+    // if there is a metadata field, add a converter for it
+    if (variantGroup.containsField("metadata")) {
+      int metadataIndex = variantGroup.getFieldIndex("metadata");
+      Type metadataField = variantGroup.getType("metadata");
+      Preconditions.checkArgument(isBinary(metadataField), "Invalid metadata field: " + metadataField);
+      converter.setConverter(metadataIndex, new VariantMetadataConverter(metadata));
+    }
 
-    // True if typed_value is an object (i.e. non-list group).
-    private boolean typedValueIsObject = false;
-    // Only used if typedValueIsObject is true.
-    private Set<String> shreddedObjectKeys;
+    return converter;
+  }
 
-    @Override
-    public VariantConverter getParent() {
-      return this.parent;
+  static ValueConverter newValueConverter(GroupType valueGroup, ParentConverter<VariantBuilder> parent) {
+    if (valueGroup.containsField("typed_value")) {
+      Type typedValueField = valueGroup.getType("typed_value");
+      if (isShreddedObject(typedValueField)) {
+        return new ShreddedObjectConverter(valueGroup, parent);
+      }
     }
 
-    @Override
-    public VariantBuilder getBuilder() {
-      return this.parent.getBuilder();
+    return new ShreddedValueConverter(valueGroup, parent);
+  }
+
+  /**
+   * Converter for metadata
+   */
+  static class VariantMetadataConverter extends BinaryConverter {
+    private final Consumer<ByteBuffer> handleMetadata;
+
+    public VariantMetadataConverter(Consumer<ByteBuffer> handleMetadata) {
+      this.handleMetadata = handleMetadata;
     }
 
-    public VariantElementConverter(VariantConverter parent, GroupType variantSchema, String objectFieldName) {
-      this(parent, variantSchema);
-      this.objectFieldName = objectFieldName;
+    @Override
+    protected void handleBinary(Binary value) {
+      handleMetadata.accept(value.toByteBuffer());
     }
+  }
 
-    public VariantElementConverter(VariantConverter parent, GroupType variantSchema) {
+  /**
+   * Converter for a non-object value field; appends the encoded value to the builder
+   */
+  static class VariantValueConverter extends BinaryConverter {
+    private final ParentConverter<VariantBuilder> parent;
+
+    public VariantValueConverter(ParentConverter<VariantBuilder> parent) {
       this.parent = parent;
-      converters = new Converter[variantSchema.getFieldCount()];
+    }
 
-      List<Type> fields = variantSchema.getFields();
+    @Override
+    protected void handleBinary(Binary value) {
+      parent.build(builder -> builder.appendEncodedValue(value.toByteBuffer()));
```

Comment:
This is only used for non-object cases. For objects, `PartiallyShreddedValueConverter` handles the `value` field by either appending the value if it is a non-object or calling `startOrContinuePartialObject` to handle the encoded base object.